hazelcast/src/main/java/com/hazelcast/collection/impl/collection/CollectionService.java

```diff
@@ -88,9 +91,25 @@ public void shutdown(boolean terminate) {
 
     @Override
     public void destroyDistributedObject(String name) {
-        CollectionContainer container = getContainerMap().remove(name);
+        final CollectionContainer container = getContainerMap().remove(name);
+        final int partitionId = nodeEngine.getPartitionService().getPartitionId(StringPartitioningStrategy.getPartitionKey(name));
         if (container != null) {
-            container.destroy();
+            if (Thread.currentThread().getClass() == PartitionOperationThread.class) {
+                container.destroy();
+            } else {
+                InternalOperationService operationService = (InternalOperationService) nodeEngine.getOperationService();
+                operationService.execute(new PartitionSpecificRunnable() {
```

Comment:
`operationService.execute` can fail in the face of migrations. We should then retry. See: https://github.com/hazelcast/hazelcast-enterprise/pull/1991

OTOH, this concurrent use/destroy is a general problem and most structures have it. I'm questioning if it makes sense to try and fix it for set/list here and leave it broken for other structures (ringbuffer, semaphore, ...) or try and devise a general solution.
See also: 
https://hazelcast.atlassian.net/wiki/spaces/PM/pages/50823201/IMap+Lifecycle+Clarification+Update